src/ORM/DataList.php

```diff
@@ -658,6 +659,39 @@ public function filterByCallback(callable $callback): SS_List
         return $output;
     }
 
+    /**
+     * Return a copy of this list which only includes items where $fieldToFilterBy matches values in $fieldInOtherList from $list.
+     *
+     * If both fields are ID, the $list dataclass must be in the same class hierarchy as the dataclass in this list
+     *
+     * @param SS_List<DataObject> $list
+     * @param string $fieldToFilterBy The name of the field in $this to filter by
+     * @param string $fieldInOtherList The name of the field in $list to get filter values from
+     * @return static<T>
+     * @throws InvalidArgumentException
+     */
+    public function filterByList(SS_List $list, string $fieldToFilterBy = 'ID', string $fieldInOtherList = 'ID'): static
```

Comment:
Using `SS_List` instead of `DataList` primarily so that we can treat a `EagerLoadedList` the same way as a `DataList` for the purposes of this method. Same as `excludeByList()`.

Since we have to widen it for that purpose, we may as well allow any `SS_List` since no additional implementation effort is required for any other list.